rke_tests/common.py

```diff
@@ -0,0 +1,214 @@
+    missing_containers = expect_containers[:]
+    for container in containers:
+        if container in expect_containers:
+            missing_containers.remove(container)
+    assert len(missing_containers) == 0, \
+        "Missing expected containers for role '{0}': {1}".format(
+            role, missing_containers)
+
+
+def validation_node_roles(nodes, k8s_nodes):
+    """
+    Validates each node's labels for match its roles
+    Validates each node's running containers match its role
+    Validates etcd etcdctl cluster-health command
+    Validates worker nodes nginx-proxy conf file for controlplane ips
+    """
+    # TODO: etcd only nodes are not listed in k8s_nodes, should handle this
+    # TODO: does hostname_override affect nginx-proxy controlplane ips?
+
+    role_matcher = {
+        'worker': 'node-role.kubernetes.io/worker',
+        'etcd': 'node-role.kubernetes.io/etcd',
+        'controlplane': 'node-role.kubernetes.io/master'}
+    controlplane = [
+        'scheduler', 'kube-controller', 'kubelet', 'kube-proxy', 'kube-api']
+    worker = ['kubelet', 'kube-proxy', 'nginx-proxy']
+    etcd = ['etcd']
+
+    controlplane_ips = []
+    for node in nodes:
+        if 'controlplane' in node.roles:
+            controlplane_ips.append(node.public_ip_address)
+
+    nodes_to_k8s_nodes = match_nodes(nodes, k8s_nodes)
+    for node, k8s_node in nodes_to_k8s_nodes:
+        for role in node.roles:
+            assert role_matcher[role] in k8s_node['metadata']['labels'].keys()
+            containers = node.docker_ps().keys()
+            if role == 'controlplane':
+                assert_containers_exist_for_role(
+                    role, containers, controlplane)
+            # nodes with work and controlplane roles do not have nginx-proxy
+            if role == 'worker' and 'controlplane' not in node.roles:
+                assert_containers_exist_for_role(role, containers, worker)
+                result = node.docker_exec(
+                    'nginx-proxy', 'cat /etc/nginx/nginx.conf')
+                for ip in controlplane_ips:
+                    assert 'server {0}:6443'.format(ip) in result, result
+            if role == 'etcd':
+                assert_containers_exist_for_role(role, containers, etcd)
+                result = node.docker_exec('etcd', 'etcdctl cluster-health')
+                assert 'cluster is healthy' in result
+
+
+def validate_private_ip_intercommunication(kubectl):
+    """
+    Creates a daemonset of pods, one pod per worker node
+    Gets pod name, pod ip, host ip, and containers
+    For each pod, use kubectl exec to ping all other pod ips
+    Asserts that each ping is successful
+    Tears down daemonset
+    """
+
+    namespace = 'default'
+    yml_file = 'resources/k8s_ymls/daemonset_test1.yaml'
+
+    # create pod on each worker node
+    result = kubectl.create_resourse_from_yml(yml_file, namespace=namespace)
+    assert result.ok, result.stderr
+
```

Comment:
Can we check the scale of the service to be the same as number of worker nodes  and make sure every worker node has 1 instance of the daemonset ?

Reply:
Split creation and validation as 2 function calls to make them usable independently.

Reply:
Parametrize namespace